### PR TITLE
Added LCA and tests

### LINKED ISSUE #62: Add algorithms for LCA in BinaryTree.
#### Description of the problem

Refer, https://github.com/codezonediitj/pydatastructs/pull/27

### PR DESCRIPTION
**Pull request by nethish**
Fixes #62 
Added algorithms for Lowest common ancestors for binary search trees.
Added tests for LCA.

### CODE REVIEW

File: pydatastructs/trees/binary_trees.py [hunk 1]
```diff
@@ -415,6 +415,107 @@ def rank(self, x):
...
+        if self._simple_path(key, self.tree[root].left, path) or \
+            self._simple_path(key, self.tree[root].right, path):
+            path.append(root)
+            return True
+
+        path.pop()
+        return False
+
+    def simple_path(self, key):
+        path = []
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: Looks like `path` is a stack. Can we use, `pydatastructs.Stack` here?

File: pydatastructs/trees/binary_trees.py [hunk 2]
```diff
@@ -415,6 +415,107 @@ def rank(self, x):
...
+        self._simple_path(key, self.root_idx, path)
+        return path
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: The logic here can be shifted to `_simple_path` as IMO, `simple_path` doesn't need to be public.

File: pydatastructs/trees/tests/test_binary_trees.py [hunk 3]
```diff
@@ -54,6 +54,13 @@ def test_BinarySearchTree():
     assert b.delete(-10) is True
     assert b.delete(-3) is True
     assert b.delete(-13) is None
+    bl = BST()
+    nodes = [50, 30, 90, 70, 100, 60, 80, 55]
+    for node in nodes:
+        bl.insert(node, node)
+    assert bl.tree[bl.lowest_common_ancestor(80, 55, 2)].key == 70
+    assert bl.tree[bl.lowest_common_ancestor(60, 70, 2)].key == 70
+    assert bl.lowest_common_ancestor(-3, 4) is None
```
Comment (czgdp1807) [inline pydatastructs/trees/tests/test_binary_trees.py]: Please add more tests. The diff coverage is low. The lines in red in [this report](https://codecov.io/gh/codezonediitj/pydatastructs/pull/88/diff) aren't executed at least once. Add tests to ensure they are executed at least once.
You can install `pytest-cov` locally. See, `.travis.yml` for commands to be used for getting the coverage report along with the tests run.

**Comment by czgdp1807** [comment]
Do not close this PR. Just push the updates in `master` branch using the remote of your work. The PR will be updated automatically.

File: pydatastructs/trees/binary_trees.py (hunk 1 above)
Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: Nope. After returning ill be iterating over the list to find first not matched item. But we cant iterate in stack.

File: pydatastructs/trees/binary_trees.py (hunk 2 above)
Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: Its true that we don't often use simple paths in BST. Are you suggesting to hide the simple_path function?

File: pydatastructs/trees/tests/test_binary_trees.py (hunk 3 above)
Comment (nethish) [reply pydatastructs/trees/tests/test_binary_trees.py]: Yes. I'll start working on that. Thanks.

File: pydatastructs/trees/binary_trees.py (hunk 2 above)
Comment (czgdp1807) [reply pydatastructs/trees/binary_trees.py]: I meant to keep all the logic for simple paths in `_simple_path`.

Comment (czgdp1807) [reply pydatastructs/trees/binary_trees.py]: Can you please tell in which lines is `simple_path` called? It is called more than once? Any research paper or Wikipedia or university lecture describing about simple paths? What are the possible use cases of simple paths other than finding LCA?

Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: I was also trying to expose the simple path function to users. That is the reason why I wrote one exposed and a hidden function. About simple path https://en.wikipedia.org/wiki/Path_(graph_theory) and CLRS book. Other use cases are to find the intermediate nodes between root and given node.

Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: Mainly used for LCA algorithm 2. So, should I remove it?

Comment (czgdp1807) [reply pydatastructs/trees/binary_trees.py]: Yes, put all the logic for finding simple paths under `_simple_path`.

File: pydatastructs/trees/binary_trees.py [hunk 4]
```diff
@@ -415,6 +415,126 @@ def rank(self, x):
...
+            u_left = self.comparator(self.tree[u].key, self.tree[curr_root].key)
+            v_left = self.comparator(self.tree[v].key, self.tree[curr_root].key)
+            while not (u_left ^ v_left):
+                if u_left and v_left:
+                    curr_root = self.tree[curr_root].left
+                else:
+                    curr_root = self.tree[curr_root].right
+                if curr_root == u or curr_root == v:
+                    return curr_root
+                u_left = self.comparator(self.tree[u].key, self.tree[curr_root].key)
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: Please split this into two lines. Doesn't look clean.

File: pydatastructs/trees/binary_trees.py [hunk 5]
```diff
@@ -415,6 +415,126 @@ def rank(self, x):
...
+
+        .. [2] https://pdfs.semanticscholar.org/e75b/386cc554214aa0ebd6bd6dbdd0e490da3739.pdf
+
+        """
+        if algorithm == 1:
+            curr_root = self.root_idx
+            u, v = self.search(j), self.search(k)
+            if (u is None) or (v is None):
+                return None
+            u_left = self.comparator(self.tree[u].key, self.tree[curr_root].key)
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: +1

File: pydatastructs/trees/binary_trees.py [hunk 6]
```diff
@@ -415,6 +415,126 @@ def rank(self, x):
...
+            v_left = self.comparator(self.tree[v].key, self.tree[curr_root].key)
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: +1

File: pydatastructs/trees/binary_trees.py [hunk 7]
```diff
@@ -415,6 +415,126 @@ def rank(self, x):
...
+            while not (u_left ^ v_left):
+                if u_left and v_left:
+                    curr_root = self.tree[curr_root].left
+                else:
+                    curr_root = self.tree[curr_root].right
+                if curr_root == u or curr_root == v:
+                    return curr_root
+                u_left = self.comparator(self.tree[u].key, self.tree[curr_root].key)
+                v_left = self.comparator(self.tree[v].key, self.tree[curr_root].key)
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: +1

File: pydatastructs/trees/binary_trees.py [hunk 8]
```diff
@@ -415,6 +415,126 @@ def rank(self, x):
...
+                    D. Harel. A linear time algorithm for the
+                    lowest common ancestors problem. In 21s
+                    Annual Symposium On Foundations of
+                    Computer Science, pages 308-319, 1980.
+
+        Returns
+        =======
+
+        int
+            The index of the lowest common ancestor in the tree.
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: IMO, returning `key` is better in case of LCA.

File: pydatastructs/trees/binary_trees.py [hunk 9]
```diff
@@ -415,6 +415,126 @@ def rank(self, x):
...
+        Parameters
+        ==========
+
+        key: Node.key
+            key of the node for which path to be found
+
+        Returns
+        =======
+
+        list
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]:
```
path: list
    [some_info_about_it]
```

File: pydatastructs/trees/binary_trees.py (hunk 8 above)
Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: Okay. Ill do that.

File: pydatastructs/trees/binary_trees.py (hunk 2 above)
Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: Finished merging.

File: pydatastructs/trees/binary_trees.py (hunk 9 above)
Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: Ok.

File: pydatastructs/trees/binary_trees.py [hunk 10]
```diff
@@ -415,6 +415,129 @@ def rank(self, x):
...
         return r
 
+    def _simple_path(self, key, root, path=[]):
+        """
+        Utility funtion to find the simple path between root and node.
+
+        Parameter
+        =========
+
+        key:
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: `key: Python object`

File: pydatastructs/trees/binary_trees.py [hunk 11]
```diff
@@ -415,6 +415,129 @@ def rank(self, x):
...
+            node to be searched
+
+        path:
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: `path: list`
Though I don't think it is needed in the API.

File: pydatastructs/trees/binary_trees.py [hunk 12]
```diff
@@ -415,6 +415,129 @@ def rank(self, x):
...
+        """
+        if root is None:
+            return False
+        path.append(root)
+        if self.tree[root].key == key:
+            return True
+
+        if self._simple_path(key, self.tree[root].left, path) or \
+            self._simple_path(key, self.tree[root].right, path):
+            return True
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: Please convert this into iterative logic using `pydatastructs.Stack` as Python has a finite recursion limit irrespective of the size of heap memory.

**Comment by czgdp1807** [comment]
@nethish Are you participating through JMOC?

File: pydatastructs/trees/binary_trees.py (hunk 12 above)
Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: Stack hasn't been implemented yet.

**Comment by czgdp1807** [comment]
@nethish Would you still like to work on it? Thanks.

**Comment by nethish** [comment]
Yes.

**Comment by czgdp1807** [comment]
@nethish Any news on this?

File: pydatastructs/trees/binary_trees.py (hunk 12 above)
Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: Yes. I made all the changes you have mentioned except this. 
pydatastruct.stack is still abstract.

Comment (czgdp1807) [reply pydatastructs/trees/binary_trees.py]: See, https://github.com/codezonediitj/pydatastructs/blob/master/pydatastructs/miscellaneous_data_structures/stack.py#L9
`Stack` can be used to remove recursion here.

Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: I just now saw that.
 Ill start making the changes

Comment (nethish) [reply pydatastructs/trees/binary_trees.py]: Done.

**Comment by nethish** [comment]
I have made all the changes.

File: pydatastructs/trees/binary_trees.py (hunk 10 above)
Comment (czgdp1807) [reply pydatastructs/trees/binary_trees.py]: This isn't addressed.

File: pydatastructs/trees/binary_trees.py [hunk 13]
```diff
@@ -415,6 +415,141 @@ def rank(self, x):
...
+        Parameter
+        =========
+
+        key:
+            key of the node to be searched
+
+        Returns
+        =======
+
+        bool:
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]:
```suggestion
        path[::-1]: bool
```

File: pydatastructs/trees/binary_trees.py [hunk 14]
```diff
@@ -415,6 +415,141 @@ def rank(self, x):
...
+                    return self.tree[curr_root].key
+
+                u_left = self.comparator(self.tree[u].key, \
+                    self.tree[curr_root].key)
+                v_left = self.comparator(self.tree[v].key, \
+                    self.tree[curr_root].key)
+
+            if curr_root is None:
+                return curr_root
+            return self.tree[curr_root].key
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: `key` may not always be an integer. So, this conflicts with the doc string.

File: pydatastructs/trees/binary_trees.py [hunk 15]
```diff
@@ -415,6 +415,141 @@ def rank(self, x):
...
+        .. [1] https://en.wikipedia.org/wiki/Lowest_common_ancestor
+
+        .. [2] https://pdfs.semanticscholar.org/e75b/386cc554214aa0ebd6bd6dbdd0e490da3739.pdf
+
+        """
+        if algorithm == 1:
+            curr_root = self.root_idx
+            u, v = self.search(j), self.search(k)
+            if (u is None) or (v is None):
+                return None
```
Comment (czgdp1807) [inline pydatastructs/trees/binary_trees.py]: A value error would have been better to let the user know that they are searching for garbage values.

**Comment by czgdp1807** [comment]
Will merge when the tests will pass.

**Pull request merged**